### ISSUE TITLE
Show "Wayland not supported" or fix it

### ISSUE THREAD

**Issue by stefonarch** [lxqt/screengrab#366]
At least until we implement https://wayland.app/protocols/wlr-screencopy-unstable-v1

##### System Information

* Distribution & Version: all

**Comment by stefonarch** [comment]
@gfgit @marcusbritanicus  Let's discuss https://github.com/lxqt/lxqt-panel/pull/2046#issuecomment-2076870391 here. 

I'm quite ignorant about portals, so no idea. It should be this one https://github.com/emersion/xdg-desktop-portal-wlr

**Comment by marcusbritanicus** [comment]
@stefonarch Thanks. I was thinking of opening an issue here, good we can continue here.

> I'm quite ignorant about portals, so no idea.

I too am not very well versed with it. Almost all my attempts to use the portals have failed. I am rather disdainful of x-d-p and it's various implementations. But I have the basic knowledge of how Wayland works, and that's sufficient for me to explain the problem.

> It should be this one https://github.com/emersion/xdg-desktop-portal-wlr

Yes, this is the one. This implements the org.freedesktop.portal.SCreeshot interface. And whatever I try, I cannot get this to work. I end up with a same error: `No such interface “org.freedesktop.portal” on object at path /org/freedesktop/portal/desktop`

The idea of these portal is simple. It was developed mainly for flatpaks. The idea is: I (the xdg-dekstop-portal or its implementations in various forms x-d-p-kde, x-d-p-lxqt, x-d-p-gtk etc) will do certain tasks for you (the client app) because you're running in a "sandbox". Eventually, most normal apps also have accepted this setup. Qt and GTK evolved to make app built using them work smoothly in flatpaks. This meant they have inbuilt support for a few certain interfaces. For example, the file dialog. You'll notice that in a well setup system, you'll almost never see the default boring Qt file open dialog. It's usually, KDE or LXQt or GTK file dialog from the portal. The communication between the app and the portal is via dbus.

All portals work nearly perfectly on X11. Why? They were designed to work on X11. Most portals work great wayland like the file-open/file-save etc. Why? Because most of these do not need any compositor interaction. 

Here comes the Screenshot protocol of x-d-p. The x-d-p process can easily access any window and its contents on X11. X11 has the concept of window ids to identify the windows. All X11 clients can access this information. So my screenshot app simply asks the Xserver to give the window id of the app that is clicked. And then it asks the x-d-p to give the screenshot. Works great!!

Enter wayland. Wayland does not have the concept of window ids. Secondly, on wayland, we treat each window and it's contents as private. A random client has no mechanism to ask the compositor to give the view under the mouse or active view or even view which was clicked. On X11/XCB we had a million functions that could do one thing or another. On wayland we can talk to the compositor via the socket, using protocols. And we do not have a protocol to query client id. So, there is no way to ask the x-d-p to get the screenshot of a window with a particular id. Secondly, there is no way for x-d-p-wlr to ask the compositor to give the screenshot of a particular window. The only protocol available is wlr-screencopy, and that allows you to ask the compositor the screenshot of the whole screen.

If you have noticed, on wlroots-based compositors, during screen-sharing, you can share only tabs (of chromium/chrome/brave/etc...) and the whole screen. You cannot share a single app surface. Plasma has its own protocol that allows sharing of particular window or tab or screen. And that is not implemented by anybody else.

**TL;DR** There is no mechanism, x-d-p, or otherwise to capture the contents of a particular window.

**Comment by stefonarch** [comment]
Afaik the only thing that looks working for me with this portal is screenrecording (obs), sharing should work too according to the[ test page ](https://mozilla.github.io/webrtc-landing/gum_test.html).

**Comment by marcusbritanicus** [comment]
That's correct. Screenshot of the whole screen. The extension of this is screen-recording/screen-casting - the protocol is capable of sending the data repeatedly.

**Comment by tsujan** [comment]
An elementary patch for anyone who wants to use screengrab under labwc or kwin_wayland. It needs `grim` under labwc and `spectacle` under kwin_wayland. Of course, its functionality is limited, and it's only a temporary patch — until the real code is written:

[wayland_screenshot.zip](https://github.com/user-attachments/files/19675089/wayland_screenshot.zip)

**Comment by marcusbritanicus** [comment]
@tsujan @marcusbritanicus FWIW.. We can quite easily write native C++ code to get screenshot for both KDE and wlroots based stuff.

**Comment by stefonarch** [comment]
We = you ? ;) Nice!

Window selection won't work though on wlroots afaik.

**Comment by marcusbritanicus** [comment]
> We = you ? ;) Nice!

He he.. Yeah I'll give you a worknig code. But I will need time to integrate it into the existing code. So.. "we" it is. I give the code. Perhaps @tsujan can insert it into the right place.

> Window selection won't work though on wlroots afaik.

No. Not yet. It will be only be the whole screen, or a selection rectangle.. Once devs of wlroots decide to include the newer wayland protocols (), then we will be able to capture windows as well.

**Comment by tsujan** [comment]
> I give the code. Perhaps @tsujan can insert it into the right place.

Yes, in general, I need to see more Qt sample codes that connect wlroots as well as kwin_wayland protocols to Qt. Then everything will become clear to me. wlroots' doc is superb, while that of Qt is poor *in this regard*. Besides, I don't learn anything by reading docs ;)

@marcusbritanicus 
Do you have any advice for me? A relatively rich source of Qt codes that connect protocols to Qt, beyond what's done in lxqt-panel?

**Comment by tsujan** [comment]
> … and search random code samples

That's what I feared.

**Comment by marcusbritanicus** [comment]
@tsujan @stefonarch Here is a very simple app that does captures the current screen. Since I am used to meson (and very bad at cmake), I have used meson. Have a look at it. If it works reasonably on wlroots based compositors, then, I'll convert ScreenShot.cpp into a class, which you can then use in the code.

https://0x0.st/8fO6.tgz

**Comment by tsujan** [comment]
It works here and is a nice start point. Thanks a lot!

I could make the required class(es), add it to the code, and probably make changes after releasing LXQt 2.2.0, but if you want to do it yourself, please tell me.

**Comment by tsujan** [comment]
But one question about the MIT License. Should it remain so or can it be changed? I know almost nothing about these things.

**Comment by tsujan** [comment]
@marcusbritanicus 
I took a look at your code and was able to find the meanings of some methods by consulting the source of QtWayland. I wonder how you found the core methods. To me, it should have been a long and difficult process of digging and testing. Am I right? Is there any hope that a documentation will be written for this part of Qt soon?

**Comment by marcusbritanicus** [comment]
> But one question about the MIT License. Should it remain so or can it be changed? I know almost nothing about these things.

It's my code. You can change it to what you like.

> I could make the required class(es), add it to the code, and probably make changes after releasing LXQt 2.2.0, but if you want to do it yourself, please tell me.

**Comment by tsujan** [comment]
> I am busy till next week…

Like me. I'll do it little by little (last night I made dirty changes to the cmake file just to start). Particularly, taking the screenshot of a region will require some layer-shell codes.

Anyway, the main job is done by your code.

> All the methods I used are from Qt's private headers.

In an ideal world, Qt should do it in `QScreen`. Most probably, that won't happen soon, if ever.

> globalAdded and everything inside it is similar to functions in WQt::Registry.

I think the extra `QEventLoop` could be removed when the code is integrated.

**Comment by tsujan** [comment]
> I also think we should use a QRubberBand to get the selection region. 

Thanks to your code, the rest is straightforward, although it might take a long time before I finish it cleanly. I used the existing "selector", with temporary changes, to test screenshots of regions, and they were OK.

In the end, I might add Wayland options for selecting the screen with multi-screen setups, because the screen that `QWindow` reports isn't reliable unless the window is completely shown, which isn't the case for the first screenshot. There are many other details too…

**Comment by tsujan** [comment]
Did it for a single screen:

![Image](https://github.com/user-attachments/assets/cfed4470-a579-4e86-a49a-d19e33a11c6c)

A good thing about Wayland is that we don't need to take a static screenshot of the whole screen for selecting a region; the region can be selected on the live screen by using a transparent overlay.

The next step will be about scaling as well as multi-screen setups. I'll make the PR after that.

**Comment by tsujan** [comment]
Good idea. Of course the tip isn't included in the screenshot, but there's no reason to keep it after the first mouse press event.

**Comment by tsujan** [comment]
I'm closing this because we started to support Wayland.

It doesn't cover compositors that don't support wlr screencopy protocol, like kwin_wayland, but that's obvious from the empty screenshots, and it won't cause any problem after https://github.com/lxqt/screengrab/pull/401
